providers/flagd/pkg/service/service.go

```diff
@@ -50,131 +76,292 @@ type resolutionResponseConstraints interface {
 		schemaV1.ResolveFloatResponse | schemaV1.ResolveObjectResponse
 }
 
-func resolve[req resolutionRequestConstraints, resp resolutionResponseConstraints](
-	ctx context.Context, logger logr.Logger,
-	resolver func(context.Context, *connect.Request[req]) (*connect.Response[resp], error),
-	flagKey string, evalCtx map[string]interface{},
-) (*resp, error) {
-	evalCtxF, err := structpb.NewStruct(evalCtx)
+func (s *Service) Init() error {
+	var err error
+	s.client, err = newClient(s.cfg)
 	if err != nil {
-		logger.Error(err, "struct from evaluation context")
-		return nil, openfeature.NewParseErrorResolutionError(err.Error())
+		return err
 	}
 
-	res, err := resolver(ctx, connect.NewRequest(&req{
-		FlagKey: flagKey,
-		Context: evalCtxF,
-	}))
-	if err != nil {
-		return nil, handleError(err)
-	}
+	ctx, cancelFunc := context.WithCancel(context.Background())
 
-	return res.Msg, nil
+	go func() {
+		s.startEventStream(ctx)
+	}()
+
+	s.cancelHook = cancelFunc
+
+	return nil
+}
+
+func (s *Service) Shutdown() {
+	s.cancelHook()
 }
 
 // ResolveBoolean handles the flag evaluation response from the flagd ResolveBoolean rpc
-func (s *Service) ResolveBoolean(ctx context.Context, flagKey string, evalCtx map[string]interface{}) (*schemaV1.ResolveBooleanResponse, error) {
-	client := s.Client.Instance()
-	if client == nil {
-		return &schemaV1.ResolveBooleanResponse{
-			Reason: string(openfeature.ErrorReason),
-		}, openfeature.NewProviderNotReadyResolutionError(ConnectionError)
+func (s *Service) ResolveBoolean(ctx context.Context, key string, defaultValue bool,
+	evalCtx map[string]interface{}) of.BoolResolutionDetail {
+
+	if s.cache.IsEnabled() {
+		fromCache, ok := s.cache.GetCache().Get(key)
+		if ok {
+			fromCacheResDetail, ok := fromCache.(openfeature.BoolResolutionDetail)
+			if ok {
+				fromCacheResDetail.Reason = ReasonCached
+				return fromCacheResDetail
+			}
+		}
```

Comment:
It would be great if we could use generics or some other mechanism to reduce this duplication.

Reply:
+1 on this. I did not attempt this to reduce review efforts. Change in the PR fully maps to current main's logic. But once merged, we might be able to find some commons and extract them to generic handlers